src/python/pants/backend/python/goals/run_python_requirement.py

```diff
@@ -0,0 +1,181 @@
+
+InvertedModuleMapping: TypeAlias = FrozenDict[Address, Tuple[str, ...]]
+
+
+def _in_chroot(relpath: str) -> str:
+    return os.path.join("{chroot}", relpath)
+
+
+@dataclass(frozen=True)
+class PythonRequirementFieldSet(RunFieldSet):
+    supports_debug_adapter = False
+    run_in_sandbox_behavior = RunInSandboxBehavior.RUN_REQUEST_HERMETIC
+
+    required_fields = (
+        PythonRequirementsField,
+        PythonRequirementDependenciesField,
+        PythonRequirementModulesField,
+        PythonRequirementTypeStubModulesField,
+        PythonRequirementResolveField,
+        PexEntryPointField,
+    )
+
+    requirements: PythonRequirementsField
+    dependencies: PythonRequirementDependenciesField
+    modules: PythonRequirementModulesField
+    resolve: PythonRequirementResolveField
+    entry_point: PexEntryPointField
+
+
+@memoized
+def _invert_module_mapping(
+    resolve: ResolveName, module_mapping: ThirdPartyPythonModuleMapping
+) -> InvertedModuleMapping:
+    """Provide an inverted module mapping that can specify the set of modules known to be fulfilled
+    by a given target address."""
+    d: dict[Address, list[str]] = defaultdict(list)
+    modules_to_providers = module_mapping.resolves_to_modules_to_providers.get(resolve)
+    if not modules_to_providers:
+        return FrozenDict()
+
+    for module_name, providers in modules_to_providers.items():
+        for provider in providers:
+            d[provider.addr].append(module_name)
+
+    return FrozenDict((address, tuple(modules)) for address, modules in d.items())
+
+
+@rule_helper
+async def _resolve_entry_point(
+    module_mapping: InvertedModuleMapping, field_set: PythonRequirementFieldSet
+) -> EntryPoint:
+
+    modules = field_set.modules.value
+    reqs = field_set.requirements.value
+    entry_point_raw = field_set.entry_point.value
+
+    if entry_point_raw:
+        resolved_entry_point = await Get(
+            ResolvedPexEntryPoint,
+            ResolvePexEntryPointRequest(field_set.entry_point),
+        )
+        entry_point = resolved_entry_point.val
+
+        if not entry_point:
+            raise Exception("The provided `entry_point` was not valid.")
+
+        return entry_point
+
+    entry_point_module: str
+    if modules and len(modules) == 1:
+        # Unambiguous module specified in the `BUILD` file
+        entry_point_module = modules[0]
+    elif len(module_mapping.get(field_set.address, ())) == 1:
+        # Check the third-party module mapping
+        entry_point_module = module_mapping[field_set.address][0]
+    elif len(reqs) == 1:
+        # Use the canonicalized project name for a single-requirement target
+        entry_point_module = canonicalize_project_name(reqs[0].project_name)
+    else:
+        raise Exception(
+            "Requirement must provide a single module, specify a single requirement, or specify "
+            "an `entry_point`"
+        )
+
+    return EntryPoint(entry_point_module)
```

Comment:
Where does the PexScript field come in? Most useful 3rdparty entry points are console scripts; so it definitely would be nice to be able to address them, although I'm not sure how you'd do this.

Reply:
Currently it doesn't!

I had thought that most useful entrypoints are those accessible by `python -m` (and that's readily supported by the current implementation), but maybe that's not a reasonable assumption to make yet?

Reply:
It may be the best you can do. Certainly, for most script runnable things like the existing tools we use, that's the expected stable entry point. The script name is a tool's brand. You can change everything else and not break CLI users, but you can't change that.